rfc8252_http_server/src/lib.rs

```diff
@@ -0,0 +1,395 @@
+) -> Result<HttpResponse> {
+    let query = req.query_string();
+    let params = match get_params(query).await {
+        Ok(params) => params,
+        Err(e) => {
+            return tokio_error(oidc_params_sender, "unknown error", &format!("{}", e)).await;
+        }
+    };
+
+    let code = params.get("code");
+    if let Some(code) = code {
+        let state = params.get("state").map(|s| s.to_string());
+        let oidc_response_params = OidcResponseParams {
+            code: code.to_string(),
+            state,
+        };
+        let _ = oidc_params_sender.send(Ok(oidc_response_params)).await;
+        accepted().await
+    } else if let Some(e) = params.get("error") {
+        if let Some(error_description) = params.get("error_description") {
+            tokio_error(oidc_params_sender, e, error_description).await
+        } else {
+            tokio_error(oidc_params_sender, e, "no error description was provided").await
+        }
+    } else {
+        tokio_error(
+            oidc_params_sender,
+            "unknown error",
+            "response parameters are missing required information",
+        )
+        .await
+    }
+}
+
+async fn accepted() -> Result<HttpResponse> {
+    Ok(HttpResponse::build(http::StatusCode::OK)
+        .content_type("text/html; charset=utf-8")
+        .body(
+            OIDCAcceptedPage {
+                product_docs_link:
+                    "https://www.mongodb.com/docs/atlas/data-federation/query/sql/drivers/odbc/connect",
+                product_docs_name: "Atlas SQL ODBC Driver",
+                error: "error",
+                error_uri: "error_uri",
+                error_description: "error_description",
+            }
+            .render()
+            .unwrap(),
+        ))
+}
+
+async fn threaded_error(
+    oidc_params_sender: mpsc::Sender<StdResult<OidcResponseParams, String>>,
+    error: &str,
+    error_description: &str,
+) -> Result<HttpResponse> {
+    let _ = oidc_params_sender.send(Err(format!("{}: {}", error, error_description)));
+    Ok(HttpResponse::build(http::StatusCode::BAD_REQUEST)
+        .content_type("text/html; charset=utf-8")
+        .body(
+            OIDCErrorPage {
+                product_docs_link: "https://www.mongodb.com/docs/atlas/data-federation/query/sql/drivers/odbc/connect",
+                product_docs_name: "Atlas SQL ODBC Driver",
+                error,
+                error_uri: "https://www.mongodb.com/docs/atlas/data-federation/query/sql/drivers/odbc/connect/bad_oidc_login",
+                error_description,
+            }
+            .render()
+            .unwrap(),
+        ))
+}
+
+async fn tokio_error(
+    oidc_params_sender: tokio_mpsc::Sender<StdResult<OidcResponseParams, String>>,
+    error: &str,
+    error_description: &str,
+) -> Result<HttpResponse> {
+    let _ = oidc_params_sender
+        .send(Err(format!("{}: {}", error, error_description)))
+        .await;
+    Ok(HttpResponse::build(http::StatusCode::OK)
```

Comment:
Should this be unauthorized (401) rather than ok (200), or some other suitable error code?

Reply:
This should have been changed with the threaded version :D